src/style-spec/function/definitions/index.js

```diff
@@ -49,46 +53,61 @@ CompoundExpression.register(expressions, {
     'pi': [ NumberType, [], () => 'Math.PI'],
     'e': [ NumberType, [], () => 'Math.E'],
     'typeof': [ StringType, [ValueType], fromContext('typeOf') ],
-    'string': defineAssertion(StringType),
-    'number': defineAssertion(NumberType),
-    'boolean': defineAssertion(BooleanType),
-    'object': defineAssertion(ObjectType),
     'to-string': [ StringType, [ValueType], fromContext('toString') ],
-    'to-number': [ NumberType, [ValueType], fromContext('toNumber') ],
-    'to-boolean': [ BooleanType, [ValueType], ([v]) => `Boolean(${v})` ],
-    'to-rgba': [ array(NumberType, 4), [ColorType], ([v]) => `${v}.value` ],
-    'to-color': [ ColorType, [ValueType], fromContext('toColor') ],
+    'to-number': [ NumberType, varargs(ValueType), (ctx, args) => {
```

Comment:
`to-number` and `to-color` now do not necessarily evaluate all their arguments -- i.e. they are special forms. Let's write them as such. I think this will also allow you to avoid all the other `ctx` and `cache` changes to this file.

Reply:
Another reason these `ctx` / `cache` changes may be useful is that they'll let us experiment with inlining arguments (rather than always caching each one as a `function eN() { ... }`)

Reply:
Moved `to-number` and `to-color` to a special form, `Coercion`.  If/when it becomes possible for `to-string` to fail, we can move it there, too.